include/swift/AST/ArgumentList.h

```diff
@@ -0,0 +1,649 @@
+
+  /// Retrieve the argument label at a given index.
+  Identifier getLabel(unsigned idx) const {
+    if (!HasLabels)
+      return Identifier();
+    return getLabelsBuffer()[idx];
+  }
+
+  /// Retrieve the argument label location at a given index.
+  SourceLoc getLabelLoc(unsigned idx) const {
+    if (!HasLabelLocs)
+      return SourceLoc();
+    return getLabelLocsBuffer()[idx];
+  }
+
+  /// Retrieve the argument at a given index.
+  Argument get(unsigned idx) const {
+    return Argument(getLabelLoc(idx), getLabel(idx), getExpr(idx));
+  }
+
+  unsigned size() const { return NumArgs; }
+  bool empty() const { return size() == 0; }
+
+  class iterator final {
+    const ArgumentList *ArgList;
+    unsigned Idx;
+
+  public:
+    using iterator_category = std::random_access_iterator_tag;
+    using value_type = Argument;
+    using pointer = const value_type *;
+    using reference = value_type;
+    using difference_type = unsigned;
+
+    iterator() : ArgList(nullptr), Idx(0) {}
+    iterator(const ArgumentList *argList, unsigned idx)
+        : ArgList(argList), Idx(idx) {}
+
+    Argument operator*() const {
+      assert(ArgList);
+      return ArgList->get(Idx);
+    }
+    Argument operator[](unsigned i) const {
+      assert(ArgList);
+      return ArgList->get(Idx + i);
+    }
+
+    friend bool operator==(const iterator &lhs, const iterator &rhs) {
+      assert(lhs.ArgList == rhs.ArgList);
+      return lhs.Idx == rhs.Idx;
+    }
+    friend bool operator!=(const iterator &lhs, const iterator &rhs) {
+      return !(lhs == rhs);
+    }
+
+    iterator &operator++() { // pre-increment
+      ++Idx;
+      return *this;
+    }
+    iterator operator++(int) { // post-increment
+      auto tmp = *this;
+      ++*this;
+      return tmp;
+    }
+    iterator &operator--() { // pre-decrement
+      --Idx;
+      return *this;
+    }
+    iterator operator--(int) { // post-decrement
+      auto tmp = *this;
+      --*this;
+      return tmp;
+    }
+
+    iterator &operator+=(unsigned i) {
+      Idx += i;
+      return *this;
+    }
+    iterator operator+(unsigned i) const {
+      return iterator(ArgList, Idx + i);
+    }
+    iterator &operator-=(unsigned i) {
+      Idx -= i;
+      return *this;
+    }
+    iterator operator-(unsigned i) const {
+      return iterator(ArgList, Idx - i);
+    }
+
+    unsigned operator-(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx - rhs.Idx;
+    }
+    bool operator<(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx < rhs.Idx;
+    }
+    bool operator<=(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx <= rhs.Idx;
+    }
+    bool operator>(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx > rhs.Idx;
+    }
+    bool operator>=(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx >= rhs.Idx;
+    }
+  };
```

Comment:
Is this basically just `llvm::indexed_accessor_iterator`? (I think this class is relatively new; there might be other places we could adopt it in Swift.)

Reply:
Ah neat! I'll go ahead and change it over to use that